docs/src/developer/typesystem.md

```diff
@@ -0,0 +1,217 @@
+in AbstractAlgebra in `src/generic/GenericTypes.jl`.
+
+The totality of all univariate polynomial types, including those of generic
+`Poly{T}` types and those coming from C libraries (such as `fmpz_poly`), is
+represented by the abstract type `PolyElem{T}` which in turn belongs to
+`RingElem`, both defined in AbstractAlgebra in `src/AbstractTypes.jl`.
+
+Similarly, the totality of all matrix types, including explicit C types
+like `fmpz_mat` and the generic `Mat{T}` types is given by the abstract type
+`MatElem{T}`, again defined in AbstractAlgebra in `src/AbstractTypes.jl`.
+
+This hierarchy of types allows one to write functions at any level, e.g. for
+all univariate polynomial types, just those with a given base type `T`, or
+for a specific concrete type corresponding to just one kind of univariate
+polynomial.
+
+A third possibility to get around the single inheritance limitation of Julia is
+type traits. There is currently no explicit compiler/language support for
+traits, however various implementations exist that make use of type parameters
+in tricky ways. This allows one to add 'traits' to types, so long as those
+traits can be expressed as types. In this way, types can have multiple
+'properties' at the same time, instead of belonging to just a single abstract
+type.
+
+Nemo does not currently use type traits, though the map types in Nemo do make
+use of a custom analogue of this.
+
+Note that unlike class based systems that dispatch on the type of the implicit
+`this` parameter, Julia methods dispatch on the type of all arguments. This is
+a natural fit for mathematics where all sorts of ad hoc left and right
+operations may be required.
+
+### Encapsulation, maps and runtime flags
+
+One limitation of the Julia approach is that the type of an object cannot be
+changed at runtime. For example one might like to insist that a given ring is
+in fact a field. There are three standard ways to handle this in Julia.
+
+The first approach is to encapsulate the object in another object which does
+have the desired type. The second approach is to map the object to a different
+one of the required type (e.g. by applying a morphism). The third approach is
+to introduce data fields in the original type which can be changed at runtime,
+unlike its type. All three approaches come with downsides. 
+
+Encapsulation can be time consuming for the developer as methods which applied
+to the original object do not automatically apply to the encapsulated object.
+One can write methods which do, but this is not automatic.
+
+Application of a map may come with a performance penalty and may be difficult
+for the user to navigate. Moreover, mutation of the resulting object does not
+result in mutation of the original object.
+
+The third option of adding runtime data fields essentially takes one back to
+writing a (possibly bug ridden) interpreter. It relies on the developer 
+implementing outer methods that make use of hand written control statements
```

Comment:
what is an "outer method"?

Reply:
No special technical definition. I simply mean a method calling other methods. We don't typically have them in Julia because they aren't needed. I don't have a better way to express this, so I'm going to leave this one, short of a suggested wording which is clearer.

Reply:
Ok, if just "method" doesn't work, keep it as is (I don't really understand this "third" option anyway, so I don't have a better suggestion).